dblookupext/miniblocksHandler.go

```diff
@@ -197,3 +197,88 @@ func (mh *miniblocksHandler) saveMiniblocksMetadata(miniblockHash []byte, minibl
 
 	return mh.miniblocksMetadataStorer.PutInEpoch(miniblockHash, buff, epoch)
 }
+
+func (mh *miniblocksHandler) blockReverted(header data.HeaderHandler, blockBody *block.Body) error {
+	headerHash, errCalculate := core.CalculateHash(mh.marshaller, mh.hasher, header)
+	if errCalculate != nil {
+		return fmt.Errorf("%w for header, header round %d", errCalculate, header.GetRound())
+	}
+
+	for index, mb := range blockBody.MiniBlocks {
+		miniblockHash, err := core.CalculateHash(mh.marshaller, mh.hasher, mb)
+		if err != nil {
+			return fmt.Errorf("%w for miniblock at index %d, header hash %x", err, index, headerHash)
+		}
+
+		err = mh.revertMiniblock(header, headerHash, mb, miniblockHash)
+		if err != nil {
+			return fmt.Errorf("%w for miniblock at index %d, mbHash %x, header hash %x", err, index, miniblockHash, headerHash)
+		}
+	}
+
+	return nil
+}
+
+func (mh *miniblocksHandler) revertMiniblock(header data.HeaderHandler, headerHash []byte, mb *block.MiniBlock, miniblockHash []byte) error {
+	miniblockHeader, err := mh.findMiniblockHandler(header, miniblockHash)
+	if err != nil {
+		return err
+	}
+
+	if miniblockHeader.GetIndexOfFirstTxProcessed() == 0 {
+		// we either revert a complete miniblock or the first partial executed miniblock header
+		err = mh.epochIndex.removeEpochByHash(miniblockHash)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = mh.removeExecutedTransactions(miniblockHeader, mb)
+	if err != nil {
+		return err
+	}
+
+	return mh.removeMiniblockMetadata(header, headerHash, miniblockHash)
+}
+
+// removeExecutedTransactions will remove only the transactions between the start and end index of the provided miniblockHeader
+func (mh *miniblocksHandler) removeExecutedTransactions(miniblockHeader data.MiniBlockHeaderHandler, mb *block.MiniBlock) error {
+	txHashes := mh.getOrderedExecutedTxHashes(miniblockHeader, mb)
+	for _, txHash := range txHashes {
+		err := mh.miniblockHashByTxHashIndexStorer.Remove(txHash)
+		if err != nil {
+			return fmt.Errorf("%w for tx hash %x", err, txHash)
+		}
+	}
+
+	return nil
+}
+
+func (mh *miniblocksHandler) removeMiniblockMetadata(
+	header data.HeaderHandler,
+	headerHash []byte,
+	miniblockHash []byte,
+) error {
+	loadedMiniblockMetadata, err := mh.loadExistingMiniblocksMetadata(miniblockHash, header.GetEpoch())
+	if err != nil {
+		return err
+	}
+
+	newMiniblocksInfo := make([]*MiniblockMetadataOnBlock, 0, len(loadedMiniblockMetadata.MiniblocksInfo))
+	for _, mbOnBlock := range loadedMiniblockMetadata.MiniblocksInfo {
+		if !bytes.Equal(mbOnBlock.HeaderHash, headerHash) {
+			newMiniblocksInfo = append(newMiniblocksInfo, mbOnBlock)
+		}
+	}
+	loadedMiniblockMetadata.MiniblocksInfo = newMiniblocksInfo
+
+	if len(loadedMiniblockMetadata.MiniblocksInfo) == 0 {
+		// we assume that the rollback is done in-sync with the node.
```

Comment:
We always revert something from the current epoch to something that might be from the current one or previous one - correct?

Reply:
absolutely. The process is like this:
`revert block <epoch e, nonce n>` 
`revert block <epoch e, nonce n-1>`
`change epoch to e-1`
`revert block <epoch e-1, nonce n-2>`
`revert block <epoch e-1, nonce n-3>`
....
the revert is always done in-sync